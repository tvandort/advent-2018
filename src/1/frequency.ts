import * as fs from 'fs';
import { promisify } from 'util';

const readFile = promisify(fs.readFile);

const INITIAL_FREQUENCY: number = 0;

export const repeatedFrequency = async (file: string): Promise<number> => {
  const data = await readFile(file, 'utf8');
  const lines = data.split('\n');
  const withoutempty = lines.filter(line => Boolean(line));
  const values = withoutempty.map(line => parseInt(line, 10));
  const previousFrequencies: { [key: string]: number } = {
    [INITIAL_FREQUENCY]: 1
  };
  let totalFrequency = INITIAL_FREQUENCY;
  let repeat: number | null = null;
  const collectFrequencies = (initialFrequency = INITIAL_FREQUENCY) => {
    totalFrequency = values.reduce((previous, current) => {
      const partialSum = current + previous;
      const partialSumCount = previousFrequencies[partialSum];
      if (!partialSumCount) {
        previousFrequencies[partialSum] = 1;
      } else {
        previousFrequencies[partialSum] += 1;
      }
      if (!repeat) {
        for (const prop of Object.keys(previousFrequencies)) {
          const count = previousFrequencies[prop];
          if (count === 2) {
            repeat = parseInt(prop, 10);
          }
        }
      }
      return partialSum;
    }, initialFrequency);

    return repeat;
  };
  let repeatFrequency = null;
  while (repeatFrequency === null) {
    repeatFrequency = collectFrequencies(totalFrequency);
  }
  return repeatFrequency;
};

export const sumFrequencies = async (file: string) => {
  const data = await readFile(file, 'utf8');
  const lines = data.split('\n');
  const withoutEmpty = lines.filter(line => Boolean(line));
  const values = withoutEmpty.map(line => parseInt(line, 10));
  const sum = values.reduce(
    (previous, current) => current + previous,
    INITIAL_FREQUENCY
  );
  return sum;
};
