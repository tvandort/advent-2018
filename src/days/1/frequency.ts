import { readCleanLines } from 'fs-util';
import { FrequencyAnalyzer } from './FrequencyAnalyzer';

const toInt = (line: string) => parseInt(line, 10);
const nextFrequency = (analyzer: FrequencyAnalyzer, value: number) => {
  analyzer.Next(value);
  return analyzer;
};

export const repeatedFrequency = async (file: string) => {
  const data = await readCleanLines(file);
  const values = data.map(toInt);

  const analyzer = new FrequencyAnalyzer();
  let index = 0;
  while (analyzer.HasRepeatedFrequency() === false) {
    analyzer.Next(values[index % values.length]);
    index++;
  }
  return analyzer.FirstRepeatedFrequency as number;
};

export const sumFrequencies = async (file: string) => {
  const data = await readCleanLines(file);
  return data.map(toInt).reduce(nextFrequency, new FrequencyAnalyzer())
    .CurrentFrequency;
};
