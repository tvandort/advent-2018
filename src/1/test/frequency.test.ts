import { join } from 'path';
import { sumFrequencies } from '../frequency';

describe('sum', () => {
  test.each([
    ['zero.txt', 0],
    ['one.txt', 1],
    ['negative-one.txt', -1],
    ['multiple-values.txt', 8]
  ])('read %p sum was %p', async (file, expectedSum) => {
    expect(await sumFrequencies(join(__dirname, file))).toBe(expectedSum);
  });

  it('throws when file does not exist', async () => {
    try {
      await sumFrequencies('does_not_exist.txt');
    } catch (error) {
      expect(error.code).toMatchSnapshot();
    }
  });
});
