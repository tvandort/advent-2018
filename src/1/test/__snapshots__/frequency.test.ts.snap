// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sum throws when file does not exist 1`] = `"ENOENT"`;
